### ISSUE TITLE
Implement `grid_sample` for sampling from a spatial tensor at arbitrary locations, similar to PyTorch

### ISSUE THREAD

**Issue by timstr** [tracel-ai/burn#2674]
### Feature description

In brief, I propose implementing [PyTorch's `grid_sample` method](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html#torch.nn.functional.grid_sample) in Burn. This method allows one to pass a tensor of points in 1D, 2D, or 3D Euclidean space and look up and interpolate values from a tensor with 1, 2, or 3 dimensions (plus features + minibatch). This differs from the existing `interpolate` method because the sample locations need not be a regular grid. This differs from the existing `gather` method because it uses floating point locations instead of indices, interpolates, and its autodiff computes gradients for the sampling locations themselves as well.

### Feature motivation

`grid_sample` is a powerful low-level primitive, closely related to the existing methods for interpolating and gathering, but with more freedom to sample values at arbitrary spatial locations and with gradient information for describing changes in the sample locations in addition to the grid values. Its possible use cases go beyond neural networks and it's well suited for more general-purpose (differentiable) tensor computation.

The tensor being sampled could be an image and the sample locations could be from a non-linear projection, such as cartesian-to-polar. The tensor being sampled could be a terrain heightmap, and the sample locations and their gradients could be used to model rainfall and erosion. Or the tensor being sampled could be a volumetric scene, and sample locations could be particles in a simulation.

I've personally used PyTorch's `grid_sample` during my thesis research project to, among other things, implement a basic but GPU parallelized renderer, implementing sphere tracing with a signed distance field represented as a volumetric tensor ([link to research](https://timstr.website/machine_learning/acoustic_reconstruction.html), 3D figures were generated this way). I would love to be able to do similar work in Rust using Burn.

**Comment by timstr** [comment]
I'm taking a stab at implementing this over at [https://github.com/timstr/burn/tree/feat/grid_sample](https://github.com/timstr/burn/tree/feat/grid_sample), so far I seem to have a working implementation of a `grid_sample_1d` op. I'll hopefully 2D and 3D grid sampling as well as autodiff working soon and will put up a pull request once I get there.

I'm quite new to Burn and it's quite a large project, so any early feedback would be very welcome!
https://github.com/timstr/burn/commit/6ebfe29d5c0c90392b6873693bf405ab2182441b

**Comment by timstr** [comment]
One thing I would like to support is different padding modes for when indices extend beyond the data, such as zero-padding, repeat-padding, and mirror-padding. In principle this choice would be well captured by a simple `enum` and would be equally valid for other convolution and padding operations in Burn which AFAICT don't yet support a choice of padding method like this. At least, I would expect to see it in [`conv2d`](https://docs.rs/burn/latest/burn/tensor/module/fn.conv2d.html) or [`ConvOptions`](https://docs.rs/burn/latest/burn/tensor/ops/struct.ConvOptions.html) but I don't see anything like zero/repeat/mirror/reflect mentioned

**Comment by dfsfdfse** [comment]
I have the same requirement here, and I tried using the slice method, but this causes my grid to keep cloning, making the loop very slow.

**Comment by laggui** [comment]
> One thing I would like to support is different padding modes for when indices extend beyond the data, such as zero-padding, repeat-padding, and mirror-padding. In principle this choice would be well captured by a simple `enum` and would be equally valid for other convolution and padding operations in Burn which AFAICT don't yet support a choice of padding method like this. At least, I would expect to see it in [`conv2d`](https://docs.rs/burn/latest/burn/tensor/module/fn.conv2d.html) or [`ConvOptions`](https://docs.rs/burn/latest/burn/tensor/ops/struct.ConvOptions.html) but I don't see anything like zero/repeat/mirror/reflect mentioned

For convolutions, anything other than zero padding is just an explicit padding operation that is applied before the convolution.

For padding, we only have `tensor.pad(padding, value)` right now which allows you to pad a tensor with an explicit value on the left, right, top, and bottom. This could be extended to other modes.

**Comment by laggui** [comment]
> I'm taking a stab at implementing this over at https://github.com/timstr/burn/tree/feat/grid_sample, so far I seem to have a working implementation of a `grid_sample_1d` op. I'll hopefully have 2D and 3D grid sampling as well as autodiff working soon and will put up a pull request once I get there.
> 
> I'm quite new to Burn and it's quite a large project, so any early feedback would be very welcome! [timstr@6ebfe29](https://github.com/timstr/burn/commit/6ebfe29d5c0c90392b6873693bf405ab2182441b)

Looks like you're on the right track!

Have you looked at the [contributor book](https://burn.dev/contributor-book/overview.html)? There are details about adding a new tensor op, maybe that could be helpful.

**Comment by dfsfdfse** [comment]
I have implemented a custom backend that supports 4-dimensional padding with zero bilinear interpolation for grid sampling. Would you mind taking a look to see if there are any other improvements that could be made? I tested it on my 4070s, and it takes 5 microseconds. Burn and CubeCL are truly impressive.
[grid-sample](https://github.com/dfsfdfse/grid-sample.git)

**Comment by nathanielsimard** [comment]
Yeah, the CubeCL kernel can be improved a lot. The max and min implementations are very suboptimal since they use branching. You could probably use select instead! Also, you're not using vectorization, so you could batch computation within a single unit. Finally, since you're doing manual bounds checking, you should use `launch_unchecked` instead of `launch`.

```rust
use cubecl::prelude::*;

#[cube(launch_unchecked)]
pub fn grid_sample_kernel<F: Float>(
    input: &Tensor<Line<F>>,
    grid: &Tensor<Line<F>>,
    output: &mut Tensor<Line<F>>,
) {
    ...
}

#[cube]
fn max<F: Float>(a: Line<F>, b: Line<F>) -> Line<F> {
    select( a > b,  a, b)
}

#[cube]
fn min<F: Float>(a: Line<F>, b: Line<F>) -> Line<F> {
    select( a < b,  a, b)
}
```

**Comment by wizenink** [comment]
Any update on this?
Maybe I can give it some time, it would be very useful to have. Any hint on where I should looks to help push this forward?

**Comment by laggui** [comment]
At a high-level, it could be added as a tensor operation similar to other utility methods in [`burn-tensor/src/tensor/grid`](https://github.com/tracel-ai/burn/tree/main/crates/burn-tensor/src/tensor/grid).

But a naive implementation will not be the most performant. So it would benefit from a custom implementation at the backend level (which means a custom CubeCL kernel to support our cuda/wgpu/metal/rocm backends). A backend extension (as linked in previous comments) is a good way to extend functionality, but for the tensor API that means adding the method to a trait that can be implemented for each backend.

**Comment by timstr** [comment]
@wizenink no update! I got pulled away into other projects. Please feel free to continue my work!

**Comment by Cielbird** [comment]
@wizenink I wrote a high level implementation of 2d sampling, building on top of @timstr's work. I followed pytorch's grid_sample for argument formats. I think it's ok to have just 2D and 3D grid sample. 3D sampling wouldn't be too different. #3495

**Comment by antimora** [comment]
Partial progress update: the forward pass (`grid_sample_2d`) is implemented in `crates/burn-tensor/src/tensor/api/float.rs`. However, there is no autodiff/backward pass yet - `crates/burn-autodiff/` contains no grid_sample implementation, so training through `grid_sample` is not currently supported. The backward pass still needs to be implemented.

**Comment by antimora** [comment]
Closing as completed. Remaining autodiff is tracked in https://github.com/tracel-ai/burn/issues/4723
